File: 이상욱/BOJ/m6w17/BOJ15649.java
package m6w17;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class BOJ15649 {
	static StringBuilder sb = new StringBuilder();

	public static void main(String[] args) throws Exception {

		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());

		int n = Integer.parseInt(st.nextToken());
		int m = Integer.parseInt(st.nextToken());

		int[] arr = new int[n];
		for (int i = 0; i < arr.length; i++) {
			arr[i] = i + 1;
		}
		int[] out = new int[n];
		boolean[] isUsed = new boolean[n];

		perm(arr, out, isUsed, 0, n, m);

		System.out.println(sb.toString());
	}

	private static void perm(int[] arr, int[] out, boolean[] isUsed, int depth, int n, int m) {

		if (depth == m) {
			for (int i = 0; i < m; i++) {
				sb.append(out[i] + " ");
			}
			sb.append("\n");
			return;
		}

		for (int i = 0; i < n; i++) {
			if (!isUsed[i]) {
				isUsed[i] = true;
				out[depth] = arr[i];
				perm(arr, out, isUsed, depth + 1, n, m);
				isUsed[i] = false;
			}
		}

	}
}

File: 이상욱/BOJ/m6w17/BOJ15649.kt
package m6w17

import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*

val sb = StringBuilder()
fun main(){

    val br = BufferedReader(InputStreamReader(System.`in`))
    var st = StringTokenizer(br.readLine())

    val N = st.nextToken().toInt()
    val M = st.nextToken().toInt()

    val arr = IntArray(N)
    for(i in arr.indices){
        arr[i] = i + 1
    }

    val output = IntArray(N)
    val isUsed = BooleanArray(N)

    perm(arr,output,isUsed,0,N,M)

    println(sb)


}

fun perm(arr: IntArray, output: IntArray, isUsed: BooleanArray, depth: Int, n: Int, m: Int) {

    if(depth == m){
        for(i in 0 until m){
            sb.append("${output[i]} ")
        }
        sb.append("\n")
        return
    }

    for(i in 0 until n){
        if(!isUsed[i]){
            isUsed[i]= true
            output[depth] = arr[i]
            perm(arr, output, isUsed, depth+1, n, m)
            isUsed[i] = false
        }
    }

}
